### PR TITLE
Add 'soft_delete_policy' to 'google_storage_bucket' resource 

### PR DESCRIPTION
**Pull request by zli82016**
Part of b/346407424

Backport of https://github.com/GoogleCloudPlatform/magic-modules/pull/10171 in provider 5.22 based on the impact to customers of the new soft delete defaults, this will allow them to opt out.

1. Cherry-picked the change to add 'soft_delete_policy' to 'google_storage_bucket' resource
2. Upgrade go api client to the version that SoftDeletePolicy field is present
3. Fixed other failed tests

Tests run on Teamcity:
[storage](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_NIGHTLYTESTS_GOOGLEBETA_PACKAGE_STORAGE/163784?hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildTestsSection=true&expandBuildDeploymentsSection=false&expandBuildChangesSection=true), [fwtransport](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_NIGHTLYTESTS_GOOGLEBETA_PACKAGE_FWTRANSPORT/163317?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildDeploymentsSection=false), [firebase](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_NIGHTLYTESTS_GOOGLEBETA_PACKAGE_FIREBASE/163782?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildDeploymentsSection=false&expandBuildChangesSection=true)

### CODE REVIEW

**Comment by rileykarson** [comment]
Just commenting on branch mechanics prior to review. I've confirmed `release-4.84.1` branch is branched from the same commit as the `v4.84.0` tag right now, and this PR will update both the changelog + code of the branch to what we intend to release.

(dupe of https://github.com/hashicorp/terraform-provider-google/pull/18395#issuecomment-2161295736)

**Comment by zli82016** [comment]
I run the tests in Teamcity for the affected services. If we prefer run test for the whole provider, feel free to let me know.

**Comment by zli82016** [comment]
Release 4.85.0 is based on 4.84.0 and no [diffs](https://github.com/hashicorp/terraform-provider-google-beta/compare/release-4.84.0...release-4.85.0?expand=1)

File: CHANGELOG.md
Comment (rileykarson) [inline CHANGELOG.md:1]:
```suggestion
## 4.85.0 (Unreleased)
```

File: CHANGELOG.md
Comment (rileykarson) [inline CHANGELOG.md:2]:
```suggestion

NOTES:
* The `4.85.0` release backports configuration for the retention period for Cloud Storage soft delete (https://cloud.google.com/resources/storage/soft-delete-announce) so that customers who have not yet upgraded to `5.22.0`+ are able to configure the retention period of objects in their buckets. By upgrading to this version and configuring or otherwise interacting with the `google_storage_bucket.soft_delete_policy` values, you will need to upgrade directly to `5.22.0`+ from `4.85.0` when upgrading to `5.X` in the future.
```

File: CHANGELOG.md
Comment (rileykarson) [reply CHANGELOG.md:7]: stale now

File: google-beta/services/firebase/resource_firebase_web_app_generated_test.go
Comment (zli82016) [reply google-beta/services/firebase/resource_firebase_web_app_generated_test.go:36]: The firebase and fwtransport tests failed when running `make test` locally. 

The resource `google_firebase_web_app` is beta only resource.

**Review by rileykarson — APPROVED** [review]
LGTM, let's get explicit approval from @SarahFrench before merging

After that we should kick off a full test run on the release branch, and ensure we release this when Sarah or a delegate is available in case anything goes wrong with the release tooling (probably tomorrow morning, but we'll work that out offline in chat)

**Comment by SarahFrench** [comment]
👋 Hey, I'm concerned about the "After that we should kick off a full test run on the release branch"

There isn't currently a good way to trigger the entire test suite. In the UI this would mean clicking through to trigger a custom build using a non-default branch in each service's build configuration. Is this limitation known about?

It could be possible to speed this up with a short script though, as custom builds can be triggered using the TeamCity API: https://www.jetbrains.com/help/teamcity/rest/start-and-cancel-builds.html#Start+Regular+Build

**Comment by zli82016** [comment]
> There isn't currently a good way to trigger the entire test suite. In the UI this would mean clicking through to trigger a custom build using a non-default branch in each service's build configuration. Is this limitation known about?
> 

Thank you for the information. I didn't realize the limitation. If we want to trigger all of the tests as we did for feature branch or the branch for functions, should we change the code in magic module?

**Comment by SarahFrench** [comment]
Are you asking if we need to make a change to the TeamCity configuration to help test the release-4.85.0 branch?

In the past we changed the TeamCity configuration so that we could test a 5.0.0 feature branch one night of the week, and all other nights we tested off main. For a long-lived feature branch I'd agree that's a good approach, but for a single test suite run of a release branch it feels a little heavy handed

**Comment by SarahFrench** [comment]
Pivoting to internal chat for discussion

**Comment by rileykarson** [comment]
Not aware of the limitation. If this'd be easy (<3 hours of manual effort) we may want to do a full run- I'll leave that decision up to the two of you- but I'd like to test a representative set of handwritten services given the client upgrade. GCS is already tested, so if we did compute + GKE + resourcemanager as well that'd be enough for me personally.

**Review by SarahFrench — APPROVED** [review]

**Pull request merged**
